File: android_native/app/src/main/java/com/example/widgetgallery/presentation/components/StepProgress.kt
package com.example.widgetgallery.presentation.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/** Horizontal multi-step indicator for wizards / checkout. */
@Composable
fun StepProgress(
    steps: List<String>,
    currentStep: Int,
    modifier: Modifier = Modifier,
) {
    val scheme = MaterialTheme.colorScheme
    Row(modifier, verticalAlignment = Alignment.Top) {
        steps.forEachIndexed { i, step ->
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                Box(
                    Modifier.size(30.dp).clip(CircleShape)
                        .background(
                            if (i <= currentStep) scheme.primary
                            else scheme.surfaceContainerHighest
                        ),
                    contentAlignment = Alignment.Center,
                ) {
                    if (i < currentStep) {
                        Icon(Icons.Filled.Check, null,
                            tint = scheme.onPrimary,
                            modifier = Modifier.size(16.dp))
                    } else {
                        Text("${i + 1}",
                            color = if (i == currentStep) scheme.onPrimary
                            else scheme.onSurfaceVariant,
                            fontWeight = FontWeight.Bold)
                    }
                }
                Spacer(Modifier.height(6.dp))
                Text(step, fontSize = 11.sp,
                    color = if (i <= currentStep) scheme.primary
                    else scheme.onSurfaceVariant,
                    fontWeight = if (i == currentStep) FontWeight.Bold
                    else FontWeight.Normal)
            }
            if (i != steps.lastIndex) {
                Box(
                    Modifier
                        .weight(1f)
                        .padding(top = 14.dp)
                        .height(2.dp)
                        .background(
                            if (i < currentStep) scheme.primary
                            else scheme.outlineVariant
                        )
                )
            }
        }
    }
}

File: lib/presentation/widgets/step_progress.dart
import 'package:flutter/material.dart';

/// A horizontal multi-step progress indicator for wizards / checkouts.
class StepProgress extends StatelessWidget {
  const StepProgress({
    super.key,
    required this.steps,
    required this.currentStep,
  });

  /// Short step labels.
  final List<String> steps;

  /// Zero-based index of the active step.
  final int currentStep;

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    return Row(
      children: [
        for (var i = 0; i < steps.length; i++) ...[
          Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              AnimatedContainer(
                duration: const Duration(milliseconds: 240),
                width: 30,
                height: 30,
                decoration: BoxDecoration(
                  shape: BoxShape.circle,
                  color: i <= currentStep
                      ? scheme.primary
                      : scheme.surfaceContainerHighest,
                ),
                child: Center(
                  child: i < currentStep
                      ? Icon(Icons.check,
                          size: 16, color: scheme.onPrimary)
                      : Text(
                          '${i + 1}',
                          style: TextStyle(
                            color: i == currentStep
                                ? scheme.onPrimary
                                : scheme.onSurfaceVariant,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                ),
              ),
              const SizedBox(height: 6),
              Text(
                steps[i],
                style: TextStyle(
                  fontSize: 11,
                  color: i <= currentStep
                      ? scheme.primary
                      : scheme.onSurfaceVariant,
                  fontWeight:
                      i == currentStep ? FontWeight.bold : FontWeight.normal,
                ),
              ),
            ],
          ),
          if (i != steps.length - 1)
            Expanded(
              child: Container(
                height: 2,
                margin: const EdgeInsets.only(bottom: 22),
                color: i < currentStep
                    ? scheme.primary
                    : scheme.outlineVariant,
              ),
            ),
        ],
      ],
    );
  }
}
